enzyme/Enzyme/TypeAnalysis/Trie.h

```diff
@@ -0,0 +1,262 @@
+class Trie : public std::enable_shared_from_this<Trie> {
+private:
+  int depth;
+  bool is_leaf;
+  std::map<int, Trie *> mapping;
+  std::optional<ConcreteType> ct;
+
+public:
+  Trie(BaseType BT, int depth = 0) {
+    depth = depth;
+    is_leaf = false;
+    ct = BT;
+    mapping = std::map<int, Trie *>{};
+  };
+  Trie() : Trie(ConcreteType(BaseType::Unknown)){};
+  Trie(ConcreteType dat, int depth = 0) {
+      depth = depth;
+      is_leaf = false;
+      ct = dat;
+      mapping = std::map<int, Trie *>{};
```

Comment:
unnecessary since already default init'd (and also constructors should use the : depth(depth), is_leaf(false) method to avoid redundant work